tenders/common/elf.c

```diff
@@ -235,3 +221,119 @@ void elf_load(const char *file, uint8_t *mem, size_t mem_size,
         close (fd_kernel);
     exit(1);
 }
+
+void elf_load_mft(const char *file, struct mft **mft, size_t *mft_size)
+{
+    int fd_kernel = -1;
+    ssize_t nbytes;
+    size_t ph_size;
+    Elf64_Off ph_off;
+    Elf64_Half ph_entsz;
+    Elf64_Half ph_cnt;
+    Elf64_Half ph_i;
+    Elf64_Phdr *phdr = NULL;
+    Elf64_Ehdr hdr;
+    struct mft *note = NULL;
+    size_t note_offset, note_size = 0;
+
+    fd_kernel = open(file, O_RDONLY);
+    if (fd_kernel == -1)
+        goto out_error;
+
+    nbytes = pread_in_full(fd_kernel, &hdr, sizeof(Elf64_Ehdr), 0);
+    if (nbytes < 0)
+        goto out_error;
+    if (nbytes != sizeof(Elf64_Ehdr))
+        goto out_invalid;
+    if (!ehdr_is_valid(&hdr))
+        goto out_invalid;
+
+    ph_off = hdr.e_phoff;
+    ph_entsz = hdr.e_phentsize;
+    ph_cnt = hdr.e_phnum;
+    ph_size = ph_entsz * ph_cnt;
+    phdr = malloc(ph_size);
+    if (!phdr)
+        goto out_error;
+    nbytes = pread_in_full(fd_kernel, phdr, ph_size, ph_off);
+    if (nbytes < 0)
+        goto out_error;
+    if (nbytes != ph_size)
+        goto out_invalid;
+
+    /*
+     * Find the NOTE containing the Solo5 manifest and sanity check its headers.
+     */
+    bool mft_note_found = false;
+    for (ph_i = 0; ph_i < ph_cnt; ph_i++) {
+        if (phdr[ph_i].p_type != PT_NOTE)
+            continue; /* Not a NOTE, next */
+
+        struct mft_note_header nhdr;
+        if (phdr[ph_i].p_filesz < sizeof nhdr)
+            continue; /* Too small to be a (valid) Solo5 NOTE, skip */
+        nbytes = pread_in_full(fd_kernel, &nhdr, sizeof nhdr,
+                phdr[ph_i].p_offset);
+        if (nbytes < 0)
+            goto out_error;
+        if (nbytes != sizeof nhdr)
+            goto out_invalid;
+        if (strncmp(nhdr.name, SOLO5_NOTE_NAME, sizeof(SOLO5_NOTE_NAME)) != 0)
+            continue; /* Not a Solo5 NOTE, next */
+
+        if (nhdr.type != SOLO5_NOTE_MANIFEST) {
+            warnx("%s: phdr[%u] contains invalid Solo5 NOTE", file, ph_i);
+            goto out_invalid;
+        }
+        if (nhdr.descsz > MFT_NOTE_MAX_SIZE) {
+            warnx("%s: phdr[%u] Solo5 NOTE size out of range", file, ph_i);
+            goto out_invalid;
+        }
+
+        mft_note_found = true;
+        break;
+    }
+    if (!mft_note_found) {
+        warnx("%s: No manifest found, not a Solo5 executable?", file);
+        goto out_invalid;
+    }
+
+    /*
+     * At this point we have verified that the NOTE at phdr[ph_i] could be a
+     * Solo5 manifest, and it's sizes are within some sane limits. Adjust size
+     * and offset to skip the note header itself and read the full note
+     * contents (the manifest) into dynamically allocated memory.
+     */
+    note_offset = phdr[ph_i].p_offset + offsetof(struct mft_note, m);
+    note_size = phdr[ph_i].p_filesz - offsetof(struct mft_note, m);
```

Comment:
At this point `phdr[ph_i].p_filesz` is guaranteed to be `>= sizeof struct mft_note_header`.
- If `phdr[ph_i].p_filesz == sizeof struct mft_note_header` you will end up with a `struct mft_note *` that points to a an allocation of 0 bytes, ie without allocated memory for the `m` member. This will later lead to a call to `mft_validate(mft_size == 0)`. If `p_filesz = 1 + sizeof struct mft_note_header` you get an `mft_size` of 1, and so on. See my comment in `mft_validate` re: dereferencing fields before validating `mft_size`.
- If there is padding between the `h` and `m` fields in `struct mft_note` this `p_filesz - offsetof()` can underflow, but that's OK since it'll fail in the malloc or in the `pread_in_full()` below.